### ISSUE TITLE
Add forks per play or per task

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#24037]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

core

##### ANSIBLE VERSION

v2.4

##### SUMMARY

We have the need to limit the number of tasks run in parallel for a specific task in order to not overload some infrastructure (in this case a terminal server). The total number of connections should not exceed 4.

Now we are using `serial:` for this, however this will do them 4 at a time, and it waits for the slowest one to be finished to do the next batch of 4. Which is suboptimal and slower, than using something like `forks: 4` where it fires off a new task as soon as one has finished.

For us this would make a lot of difference.

**Comment by dagwieers** [comment]
Another reason for being able to do this per task is that it makes it possible to be used in a role, which is not possible with `serial:`. Usually there's only one task that has this requirement (possibly one that uses delegate_to by necessity, like a module used to talk to appliances).

@mpdehaan actually hinted at this possibility in 2014: https://groups.google.com/d/msg/ansible-project/rBcWzXjt-Xc/_QCTljBcCG0J

**Comment by infrascripting** [comment]
+1

I have an instance where I am hammering a RedHat Satellite server to enable various repos for the client machines. This is very resource-intensive on the Satellite server if they all connect at once. At any point if >15 servers are trying to enable 5-8 repos at a time, the chance that one or more will fail is close to 100%, due to the sheer amount of traffic to the Satellite server.

Best-case scenario right now is that I serialize an entire role in the middle of my playbook in order to scale back the number of connections at a time. This is frustrating as it is only that one task that would necessitate a throttling of network traffic.

**Comment by signal64** [comment]
+1

We have multiple high latency + low bandwidth connections over c-band satellite where at times during the play we need to ensure we aren't saturating the connections (eg. pushing files). As mentioned serial waits for all hosts in the batch to complete before moving onto the next group of hosts. Connections that are slower than the rest will hold up the play and not move onto other available forks.

**Comment by codecircuit** [comment]
+1

In our infrastructure the nodes need to register at our IPA-Server, which can not handle a group of nodes simultaneously. With the `serial` option on a task-level, we could simply
serialize the registration task.

**Comment by grizmin** [comment]
+1
If you need to restart member nodes of a cluster 1 by 1 there is no simple solution. Forks per task will be a easy and simple solution of the problem.

**Comment by Bak3y** [comment]
I have a need to set this on a playbook basis right now.  There's certain times I'd love to control this inside the playbook itself rather than relying on --forks

**Comment by orlra** [comment]
In my scenario I have playbook with light/heavy workloads in 1 playbook, 
if I could:
include: light_workload.yml
with_forks: 100%
include: heavy_workload.yml
with_forks: 25%
that would certainly save me some optimization work.
I'm open for any per-task throttling option.

**Comment by mhalano** [comment]
I think you mean "serial" per play or per task. I ran some tests and the option "forks" is a very low level option which use to avoid overload on controller node. I ran a play with "serial = 3" and "forks = 2" and the task ran on 3 hosts, just slowly because Ansible waited one fork to finish to start another fork.

**Comment by dagwieers** [comment]
@mhalano No, I definitely mean **forks** as **serial** has a different behaviour where it processes targets in a batch before continuing with the next batch. This was explained above.

**Comment by dkarakas1** [comment]
+1 for all the reasons mentioned above.  It sucks to have to make a playbook run at the speed of the slowest task.

**Comment by backbord** [comment]
these might be of interest: #39374 and associated PR #39602

**Comment by crossan007** [comment]
Another use case for this is the case of "multiple ansible hosts per server" when multiple instances of an application are defined on the same `ansible_host` for multiple inventory entries.

The ability to restrict certain portions of a play to a single "instance" at a time (such as configuring IIS), while allowing other portions of the play to continue in parallel would be excellent.

**Comment by bcoca** [comment]
Most of these uses cases are already covered, I know @dagwieers has a couple that are not, but so people don't wait around, I'll give some examples below:

@dkarakas1 you might want to look into 'free' strategy or 'host_pinned' as they deal avoiding  'slow tasks/hosts' as bottlenecks.

@grizmin  @crossan007 that is actually simple to do with multiple plays per playbook or with run once + loop
```yaml
- name: unrestricted
  hosts: all
  tasks:
   ...

- name: only one
  hosts: all
  serial: 1
  tasks:
    ....
  
- name: unrestricted
  hosts: all
  tasks:
   ...
```
or 

```yaml
 - name: restart each serially
   delegate_to:  '{{item}}'
   action: ....
   run_once: true
   loop: '{{ groups["all"]}}'
```

**Comment by crossan007** [comment]
Also of interest #12170

**Comment by bcoca** [comment]
@crossan007 https://github.com/ansible/ansible/issues/12170#issuecomment-258482975

**Comment by dagwieers** [comment]
Anyone is invited to test #42528 for their use-cases, and add a :+1: to the PR if you approve.

**Comment by bcoca** [comment]
fixed via #60702
